### DISCUSSION TITLE
TypeError: expected str, bytes or os.PathLike object, not Image

### DISCUSSION THREAD

**Discussion by SiYeon-Kwon** [Project-MONAI/MONAI#5447] [General]
Hi, I am using MONAI tutorial.
I'm making a new code by combining the codes of tutorial and Kaggle, but there's been a problem recently.

I googled the problem, but I haven't found the answer yet. Can you help me?

Thank you to everyone.

//
Exception has occurred: RuntimeError
Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/kwon/my/MONAI/monai/transforms/transform.py", line 102, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "/home/kwon/my/MONAI/monai/transforms/transform.py", line 66, in _apply_transform
    return transform(parameters)
  File "/home/kwon/my/MONAI/monai/transforms/io/array.py", line 231, in __call__
    filename = tuple(f"{Path(s).expanduser()}" for s in ensure_tuple(filename))  # allow Path objects
  File "/home/kwon/my/MONAI/monai/transforms/io/array.py", line 231, in <genexpr>
    filename = tuple(f"{Path(s).expanduser()}" for s in ensure_tuple(filename))  # allow Path objects
  File "/usr/lib/python3.10/pathlib.py", line 960, in __new__
    self = cls._from_parts(args)
  File "/usr/lib/python3.10/pathlib.py", line 594, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/usr/lib/python3.10/pathlib.py", line 578, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not Image

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/kwon/.local/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 302, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/kwon/.local/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 58, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/kwon/.local/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 58, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/kwon/my/MONAI/brain_tumor_monai.py", line 94, in __getitem__
    img = self.transform(img)
  File "/home/kwon/my/MONAI/monai/transforms/compose.py", line 174, in __call__
    input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
  File "/home/kwon/my/MONAI/monai/transforms/transform.py", line 129, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.io.array.LoadImage object at 0x7f9bc2bda6b0>
  File "/home/kwon/my/MONAI/brain_tumor_monai.py", line 149, in <module>
    for batch_data in train_loader:
//

**Comment by KumoLiu** [comment]
Hi @SiYeon-Kwon, it seems you passed `Image` to `LoadImage` instead of the filename.
https://github.com/Project-MONAI/MONAI/blob/ef68f16dde33172b2eeae5397ac75aa717808875/monai/transforms/io/array.py#L222-L227
Hope it can help you, thanks!

**Comment by SiYeon-Kwon** [comment]
> Hi @SiYeon-Kwon, it seems you passed `Image` to `LoadImage` instead of the filename.
> 
> https://github.com/Project-MONAI/MONAI/blob/ef68f16dde33172b2eeae5397ac75aa717808875/monai/transforms/io/array.py#L222-L227
> 
> Hope it can help you, thanks!

I tried printing the type of my code.
I did that, and I got a class('dict') type, what's the problem?
plz tell me @KumoLiu

**Comment by SiYeon-Kwon** [comment]
I think MONAI's own transform doesn't fit other codes.
I tried using the transform provided by pytorch, and another error appeared.

**Comment by KumoLiu** [comment]
Hi @SiYeon-Kwon, could you please give more details, you print which variable's type?

**Comment by SiYeon-Kwon** [comment]
@KumoLiu Can I show you the code?

**Comment by KumoLiu** [comment]
> @KumoLiu Can I show you the code?

Of course, if you could.

**Comment by SiYeon-Kwon** [comment]
//
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import cv2
import torch
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms, utils
from PIL import Image

import torchvision
from torchvision import datasets, models, transforms

from timeit import default_timer as timer

from monai.transforms import (
    Activations,
    EnsureChannelFirst,
    AsDiscrete,
    Compose,
    LoadImage,
    RandFlip,
    RandRotate,
    RandZoom,
    ScaleIntensity,
    ToTensor,
)
from monai.networks.nets import DenseNet121
from monai.metrics import ROCAUCMetric

```
classes = ['glioma', 'meningioma', 'notumor', 'pituitary']
num_class = len(classes)

def data_count(main_path = r'/home/kwon/my/MONAI/archive/Training'):
    counts = {}
    for root,dirs,files in os.walk(main_path):
        if files != [] and str(root.split('/')[-1]) in classes:
            counts[str(root.split('/')[-1])] = len(files)
    return counts

counts = data_count()
plt.bar(list(counts.keys()),list(counts.values()))
#plt.show()

def data_split(main_path = r'/home/kwon/my/MONAI/archive/Training', split_percentage = 0.3):
    train_dict = {}
    valid_dict = {}
    counts = data_count(main_path)
    for root,dirs,files in os.walk(main_path):
        if files != [] and str(root.split('/')[-1]) in ['glioma', 'meningioma', 'notumor', 'pituitary']:
            file_paths = [os.path.join(root,files[i]) for i in range(len(files))]
            
            valid_idxs = np.random.randint(low=0,high=len(files),size=int(len(files)*split_percentage))
            train_idxs = list(set(range(0,len(files))) - set(valid_idxs))
            
            train_dict[str(root.split('/')[-1])] = [file_paths[idx] for idx in train_idxs]
            valid_dict[str(root.split('/')[-1])] = [file_paths[idx] for idx in valid_idxs]
            
    return train_dict, valid_dict

train_dict, valid_dict = data_split()

print(type(train_dict))
print(type(valid_dict))

print('training data sizes : ', [len(l) for l in train_dict.values()])
print('validation data sizes : ', [len(l) for l in valid_dict.values()])

class MRIData(Dataset):
    """Chess Piece Dataset class"""
    
    def __init__(self, data_dict, transform=None):
        """
        Args:
            data_dict (dict): dictionary with class as key and the corresponding paths to the data of that class 
        """
        self.data_dict = data_dict
        self.transform = transform
        
    def __len__(self):
        return sum([len(l) for l in self.data_dict.values()])
    
    def __getitem__(self,idx):
        counts = [len(l) for l in self.data_dict.values()]
        sum_counts = list(np.cumsum(counts))
        sum_counts = [0] + sum_counts + [np.inf]
        
        for c,v in enumerate(sum_counts):
            if idx < v:
                i = (idx - sum_counts[c-1]) - 1
                break
        #print(c,i,idx)
        label = list(self.data_dict.keys())[c-1]
        img = Image.open(self.data_dict[str(label)][i]).convert('RGB')
        if self.transform:
            img = self.transform(img)
            
        return img, classes.index(str(label))

train_transforms = Compose([
    LoadImage(image_only=True),
    EnsureChannelFirst(),
    ScaleIntensity(),
    RandRotate(range_x=15, prob=0.5, keep_size=True),
    RandFlip(spatial_axis=0, prob=0.5),
    RandZoom(min_zoom=0.9, max_zoom=1.1, prob=0.5, keep_size=True),
    ToTensor()
])

val_transforms = Compose([
    LoadImage(image_only=True),
    EnsureChannelFirst(),
    ScaleIntensity(),
    ToTensor()
])

act = Activations(softmax=True)
to_onehot = AsDiscrete(to_onehot=num_class)

train_ds = MRIData(train_dict, train_transforms)
val_ds = MRIData(valid_dict, val_transforms)
test_ds = MRIData(valid_dict, val_transforms)

train_loader = DataLoader(train_ds, batch_size=64, shuffle=True, num_workers=2)
valid_loader = DataLoader(val_ds, batch_size=64, num_workers=2)
test_loader = DataLoader(test_ds, batch_size=64, num_workers=2)

for i_batch, sample_batched in enumerate(train_loader):
    print(i_batch, sample_batched[0].size(),sample_batched[1])
    print(type(sample_batched[0]))
    break
//
```

@KumoLiu  Thanks a lot KumoLiu, everytime!!!

**Comment by KumoLiu** [comment]
Hi @SiYeon-Kwon, you load the image in `Dataset`, remove this line in Dataset
```
img = Image.open(self.data_dict[str(label)][i]).convert('RGB')
```
or remove `LoadImage` in transforms.
```
LoadImage(image_only=True),
``` 
Thanks!

**Comment by SiYeon-Kwon** [comment]
Hi, @KumoLiu 
I deleted img = Image.open(self.data_dict[str(label)][i]).convert('RGB') and then, I made the variable img into a global variable.
When I set the img as a regional variable, there was a problem that it was used before allocation.
But the problem is, if I do this, I get an error saying that img is not defined.
What should I do?

**Comment by KumoLiu** [comment]
Hi @SiYeon-Kwon, you could apply transforms like this.
https://github.com/Project-MONAI/MONAI/blob/a9f9987fbc35e0088e7cd6f1536c35275e091806/monai/data/dataset.py#L85-L90

And I will transfer this issue to discussions, thanks!

**Comment by SiYeon-Kwon** [comment]
In my opinion, I don't think deleting Image.open or LoadImage will solve the problem.
I think the value returned by MRIData class does not match the transforms.
What should I do?
When I already returned the img on MRIData, I tried to return it through the index, but there was no change.
Please help me......
@KumoLiu

**Reply by KumoLiu** [reply]
Hi @SiYeon-Kwon, you can remove `LoadImage(image_only=True)` this line in your transform then you don't need to change your dataset. Since the problem "TypeError: expected str, bytes or os.PathLike object, not Image" is that you have used PIL to load the image, the output will turn to the array, then you can't pass it to LoadImage.
https://github.com/Project-MONAI/MONAI/blob/a9f9987fbc35e0088e7cd6f1536c35275e091806/monai/transforms/io/array.py#L222-L227
Hope I have explained it clearly, thanks!

**Reply by SiYeon-Kwon** [reply]
Hi, @KumoLiu!!
I modified my code.
I made the img into an array using np.array.
However, there is a new error in doing so.
'stack expects each tensor to be equal size, but got [1, 198, 150, 3] at entry 0 and [1, 512, 512, 3] at entry 1'

So I resize (512) in composite to adjust the size, but if I do that, there will be another error.
'applying transform <monai.transforms.spatial.array.Resize object at 0x0000023D8969A280>'

What should I do?
Save me from infinite errors.

**Reply by KumoLiu** [reply]
Hi @SiYeon-Kwon, I think you may need to ensure that after all the transforms, all your data need to be in the same shape. Thanks!

**Reply by SiYeon-Kwon** [reply]
I checked all my data has 512 size.

if I resize 512, error is reveal.
'applying transform <monai.transforms.spatial.array.Resize object at 0x0000023D8969A280>'

What should I do?

**Reply by KumoLiu** [reply]
Hi @SiYeon-Kwon, from the information you posted, I couldn't see what caused the error, could you show the whole error message?
Thanks!
